util/bloom/xor.go

```diff
@@ -93,28 +94,37 @@ func (xf *XorFilter) MarshalBinary() ([]byte, error) {
 // ErrBadBinary is returned when UnmarshalBinary fails
 var ErrBadBinary = errors.New("bad XorFilter binary")
 
+// TODO: make this an option to UnmarshalBinary, or a settable global, ...
+const maxFingerprints = 1000000
+
 // UnmarshalBinary implements encoding.BinaryUnmarshaller interface
 func (xf *XorFilter) UnmarshalBinary(data []byte) error {
 	pos := 0
 	var dp int
 	xor := new(xorfilter.Xor32)
 	xor.Seed, dp = binary.Uvarint(data[pos:])
-	if dp < 0 {
+	if dp <= 0 {
 		return ErrBadBinary
 	}
 	pos += dp
 	blockLength, dp := binary.Uvarint(data[pos:])
-	if dp < 0 {
+	if dp <= 0 {
 		return ErrBadBinary
 	}
 	xor.BlockLength = uint32(blockLength)
 	pos += dp
 	lenFingerprints, dp := binary.Uvarint(data[pos:])
-	if dp < 0 {
+	if dp <= 0 {
 		return ErrBadBinary
 	}
 	pos += dp
 	if lenFingerprints > 0 {
+		if lenFingerprints > maxFingerprints {
+			return ErrBadBinary
+		}
+		if (lenFingerprints * 4) > uint64(len(data)-pos) {
```

Comment:
instead of 4, use `unsafe. SizeUint32`

Reply:
seems strange to refer to `unsafe` for a universal constant ("int32 is 4 bytes")

Reply:
If you can find other location that makes the same constant - that would work as well.